### ISSUE TITLE
[BUG] The master and stable branch ISOs can't provision a cluster occasionally

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#6662]
**Describe the bug**

We noticed sometimes the master/stable branch ISO can't provision a cluster. Rancher complains the Harvester version is not included in the cluster repo:

For example:
```
2024/09/25 06:10:16 [ERROR] error syncing 'fleet-local/harvester': handler mcc-bundle: no chart version found for harvester-0.0.0-master-dd05de76, requeuing
2024/09/25 06:10:16 [ERROR] error syncing 'fleet-local/harvester-crd': handler mcc-bundle: no chart version found for harvester-crd-0.0.0-master-dd05de76, requeuing
2024/09/25 06:10:17 [ERROR] error syncing 'fleet-local/harvester-crd': handler mcc-bundle: no chart version found for harvester-crd-0.0.0-master-dd05de76, requeuing
```

And if we check the cluster-repo, the version is not there:
```
harvester-cluster-repo-6c6458bd46-pgrdg:/srv/www/htdocs/charts # grep harvester-crd-0.0.0 *
index.yaml:    - harvester-crd-0.0.0-master-9a146fc9.tgz
harvester-cluster-repo-6c6458bd46-pgrdg:/srv/www/htdocs/charts # ls 
harvester-0.0.0-master-9a146fc9.tgz	   harvester-vm-import-controller-0.4.0.tgz  rancher-logging-crd-103.1.0+up4.4.0.tgz
harvester-crd-0.0.0-master-9a146fc9.tgz    index.yaml				     rancher-monitoring-103.1.1+up45.31.1.tgz
harvester-pcidevices-controller-0.4.0.tgz  nvidia-driver-runtime-0.4.0.tgz	     rancher-monitoring-crd-103.1.1+up45.31.1.tgz
harvester-seeder-0.4.0.tgz		   rancher-logging-103.1.0+up4.4.0.tgz
```

cc @tserong @FrankYang0529 

**To Reproduce**
1. Download a master or branch ISO and provision a cluster with ipxe-examples.

**Expected behavior**

The cluster should be provisioned fine.

**Support bundle**

**Environment**
 - Harvester ISO version: 
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

- It seems to happen easily when we merge multiple PR to a branch in a short period and CI jobs are run together.
- The workaround is to trigger the CI jobs on the latest commit and rebuild a new ISO.

**Comment by ibrokethecloud** [comment]
i suspect this is because of the fact that iso build is a separate job in the action workflow

https://github.com/actions/checkout/issues/19

the code check does not persist across jobs so one option would be to staged checked out code as an artifact and download it for subsequent jobs

**Comment by tserong** [comment]
JFTR, when I hit this, these are the images I had:

```
> ls -l *broken
-rw-r--r-- 1 tserong users 6187974656 Sep 25 14:58 harvester-master-amd64.iso.broken
-rw-r--r-- 1 tserong users   96000436 Sep 25 14:55 harvester-master-initrd-amd64.broken
-rw-r--r-- 1 tserong users  806096896 Sep 25 14:55 harvester-master-rootfs-amd64.squashfs.broken
-rw-r--r-- 1 tserong users   11670976 Sep 25 14:55 harvester-master-vmlinuz-amd64.broken

> sudo mount -o loop harvester-master-amd64.iso.broken /mnt ; head -n1 /mnt/harvester-release.yaml ; sudo umount /mnt
mount: /mnt: WARNING: source write-protected, mounted read-only.
harvester: 099cec50

> sudo mount -o loop harvester-master-rootfs-amd64.squashfs.broken /mnt ; head -n1 /mnt/etc/harvester-release.yaml ; sudo umount /mnt
harvester: dd05de76
```

So the ISO and the squashfs were built/published at very close to the the same time, but were one commit off each other (ISO at 099cec50 is one commit before squashfs at dd05de76f)

**Comment by ibrokethecloud** [comment]
yeah.. code checkout happens again for build-iso since its a different job, so commit changes since there has been another merge since when the CI run started and we got to the build-iso job

**Comment by tserong** [comment]
Ah, gotcha.

**Comment by tserong** [comment]
Hang on a sec. I get that currently we checkout the code four times (once in build-binaries, once in build-push-images, once in manifest-images, once in build-iso), and that if there's new commits that collide with a CI job inflight, we could end up (e.g.) building or pushing older harvester container images, then building a newer ISO than expected. And I can see how #6703 and later #6708 address that potential problem. But, I don't think that's the problem I hit. My problem was a discrepancy between the commit of the ISO image, and the commit of the squashfs. Those two things are both built in the one job (build-iso), so AFAIK the only way I could have ended up with that discrepancy is if two separate CI runs were racing on the upload-iso step of the build-iso job.

**Comment by tserong** [comment]
JFTR, further attempts were https://github.com/harvester/harvester/pull/6713 and https://github.com/harvester/harvester/pull/6714. These were ultimately unsuccessful and were rolled back along with #6682 and #6708 in https://github.com/harvester/harvester/pull/6715.

**Comment by bk201** [comment]
Reopen, we still have the race.
